File: docs/js/kalshi-client.js
/**
 * Kalshi API Client — runs entirely in the browser.
 * Public endpoints need no auth. Authenticated endpoints use Web Crypto API
 * to sign requests with RSA-PSS (same algorithm as the Python backend).
 */

const KALSHI_BASE = 'https://api.elections.kalshi.com/trade-api/v2';
const KALSHI_DEMO = 'https://demo-api.kalshi.co/trade-api/v2';

export class KalshiClient {
  constructor() {
    this.apiKeyId = '';
    this._privateKey = null;   // CryptoKey object
    this._privateKeyPem = '';
    this.useDemo = true;
    this.authenticated = false;
  }

  get baseUrl() {
    return this.useDemo ? KALSHI_DEMO : KALSHI_BASE;
  }

  // ── Auth Setup ──────────────────────────────────────────────────

  async setCredentials(apiKeyId, privateKeyPem, useDemo = true) {
    this.apiKeyId = apiKeyId;
    this._privateKeyPem = privateKeyPem;
    this.useDemo = useDemo;
    this._privateKey = await this._importKey(privateKeyPem);
    this.authenticated = true;
  }

  async _importKey(pem) {
    const pemBody = pem
      .replace(/-----BEGIN (RSA )?PRIVATE KEY-----/, '')
      .replace(/-----END (RSA )?PRIVATE KEY-----/, '')
      .replace(/\s/g, '');
    const binaryDer = Uint8Array.from(atob(pemBody), c => c.charCodeAt(0));
    return crypto.subtle.importKey(
      'pkcs8',
      binaryDer.buffer,
      { name: 'RSA-PSS', hash: 'SHA-256' },
      false,
      ['sign']
    );
  }

  async _sign(message) {
    const enc = new TextEncoder().encode(message);
    const sig = await crypto.subtle.sign(
      { name: 'RSA-PSS', saltLength: 32 },
      this._privateKey,
      enc
    );
    return btoa(String.fromCharCode(...new Uint8Array(sig)));
  }

  _authHeaders(method, path) {
    // Returns a promise because signing is async
    const ts = Date.now().toString();
    const pathNoQuery = path.split('?')[0];
    return this._sign(ts + method.toUpperCase() + pathNoQuery).then(sig => ({
      'KALSHI-ACCESS-KEY': this.apiKeyId,
      'KALSHI-ACCESS-SIGNATURE': sig,
      'KALSHI-ACCESS-TIMESTAMP': ts,
    }));
  }

  async _get(path, params = {}, auth = false) {
    const qs = new URLSearchParams(params).toString();
    const fullPath = qs ? `${path}?${qs}` : path;
    const url = this.baseUrl + fullPath;
    const headers = auth && this.authenticated ? await this._authHeaders('GET', path) : {};
    const res = await fetch(url, { headers });
    if (!res.ok) throw new Error(`Kalshi API ${res.status}: ${await res.text()}`);
    return res.json();
  }

  async _post(path, data) {
    const url = this.baseUrl + path;
    const headers = {
      'Content-Type': 'application/json',
      ...(this.authenticated ? await this._authHeaders('POST', path) : {}),
    };
    const res = await fetch(url, { method: 'POST', headers, body: JSON.stringify(data) });
    if (!res.ok) throw new Error(`Kalshi API ${res.status}: ${await res.text()}`);
    return res.json();
  }

  async _delete(path) {
    const url = this.baseUrl + path;
    const headers = this.authenticated ? await this._authHeaders('DELETE', path) : {};
    const res = await fetch(url, { method: 'DELETE', headers });
    if (!res.ok) throw new Error(`Kalshi API ${res.status}: ${await res.text()}`);
    return res.json();
  }

  // ── Public Market Data (no auth) ───────────────────────────────

  async getEvents(params = {}) {
    const defaults = { status: 'open', limit: 200, with_nested_markets: 'true' };
    return this._get('/events', { ...defaults, ...params });
  }

  async getEvent(ticker) {
    return this._get(`/events/${ticker}`, { with_nested_markets: 'true' });
  }

  async getMarkets(params = {}) {
    const defaults = { status: 'open', limit: 200 };
    return this._get('/markets', { ...defaults, ...params });
  }

  async getMarket(ticker) {
    return this._get(`/markets/${ticker}`);
  }

  async getOrderbook(ticker) {
    return this._get(`/markets/${ticker}/orderbook`);
  }

  // ── Sports Market Discovery ────────────────────────────────────

  async findSportsMarkets(keywords) {
    const allMarkets = [];
    let cursor = null;
    let pages = 0;
    do {
      const params = { status: 'open', limit: 200, with_nested_markets: 'true' };
      if (cursor) params.cursor = cursor;
      const data = await this._get('/events', params);
      const events = data.events || [];
      for (const event of events) {
        const combined = `${event.title || ''} ${event.sub_title || ''} ${event.category || ''}`.toLowerCase();
        if (keywords.some(kw => combined.includes(kw.toLowerCase()))) {
          for (const m of (event.markets || [])) {
            m._event_title = event.title || '';
            m._event_category = event.category || '';
            allMarkets.push(m);
          }
        }
      }
      cursor = data.cursor;
      pages++;
    } while (cursor && pages < 5);
    return allMarkets;
  }

  async findTennisMarkets() {
    return this.findSportsMarkets([
      'tennis', 'ATP', 'WTA', 'Grand Slam', 'Roland Garros',
      'Wimbledon', 'US Open Tennis', 'Australian Open Tennis',
      'French Open',
    ]);
  }

  async findBaseballMarkets() {
    return this.findSportsMarkets([
      'baseball', 'MLB', 'World Series', 'American League',
      'National League', 'Yankees', 'Dodgers', 'Mets', 'Red Sox',
      'Cubs', 'Astros', 'Braves', 'Phillies', 'Padres',
    ]);
  }

  // ── Portfolio (auth required) ──────────────────────────────────

  async getBalance() {
    return this._get('/portfolio/balance', {}, true);
  }

  async getPositions() {
    return this._get('/portfolio/positions', { settlement_status: 'unsettled' }, true);
  }

  async getFills(limit = 100) {
    return this._get('/portfolio/fills', { limit }, true);
  }

  // ── Orders (auth required) ─────────────────────────────────────

  async createOrder(ticker, side, action, count, priceCents, timeInForce = 'gtc') {
    const data = {
      ticker,
      side,
      action,
      count,
      type: 'limit',
      client_order_id: crypto.randomUUID(),
      time_in_force: timeInForce,
    };
    if (side === 'yes') data.yes_price = priceCents;
    else data.no_price = priceCents;
    return this._post('/portfolio/orders', data);
  }

  async cancelOrder(orderId) {
    return this._delete(`/portfolio/orders/${orderId}`);
  }

  async getOrders(status = 'resting') {
    return this._get('/portfolio/orders', { status }, true);
  }
}

File: backend/kalshi_client.py
"""
Kalshi API Client — handles authentication, market discovery, and order execution.
Uses RSA-PSS signing per Kalshi's API spec.
"""
import base64
import datetime
import json
import time
import uuid
import logging
from typing import Any, Dict, List, Optional
from pathlib import Path

import requests
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding, rsa
from cryptography.hazmat.backends import default_backend

logger = logging.getLogger(__name__)


class KalshiClient:
    """Authenticated client for the Kalshi trading API."""

    def __init__(self, api_key_id: str, private_key_path: str, base_url: str):
        self.api_key_id = api_key_id
        self.base_url = base_url.rstrip("/")
        self._private_key = self._load_private_key(private_key_path) if private_key_path else None
        self._session = requests.Session()
        self._session.headers.update({"Content-Type": "application/json"})

    def _load_private_key(self, path: str) -> rsa.RSAPrivateKey:
        with open(path, "rb") as f:
            return serialization.load_pem_private_key(
                f.read(), password=None, backend=default_backend()
            )

    def _sign(self, message: str) -> str:
        signature = self._private_key.sign(
            message.encode("utf-8"),
            padding.PSS(
                mgf=padding.MGF1(hashes.SHA256()),
                salt_length=padding.PSS.DIGEST_LENGTH,
            ),
            hashes.SHA256(),
        )
        return base64.b64encode(signature).decode("utf-8")

    def _auth_headers(self, method: str, path: str) -> Dict[str, str]:
        ts = str(int(datetime.datetime.now().timestamp() * 1000))
        path_no_query = path.split("?")[0]
        sig = self._sign(ts + method.upper() + path_no_query)
        return {
            "KALSHI-ACCESS-KEY": self.api_key_id,
            "KALSHI-ACCESS-SIGNATURE": sig,
            "KALSHI-ACCESS-TIMESTAMP": ts,
        }

    def _get(self, path: str, params: Optional[Dict] = None, auth: bool = False) -> Dict:
        url = self.base_url + path
        headers = self._auth_headers("GET", path) if auth and self._private_key else {}
        resp = self._session.get(url, headers=headers, params=params, timeout=15)
        resp.raise_for_status()
        return resp.json()

    def _post(self, path: str, data: Dict, auth: bool = True) -> Dict:
        url = self.base_url + path
        headers = self._auth_headers("POST", path) if auth and self._private_key else {}
        resp = self._session.post(url, headers=headers, json=data, timeout=15)
        resp.raise_for_status()
        return resp.json()

    def _delete(self, path: str, auth: bool = True) -> Dict:
        url = self.base_url + path
        headers = self._auth_headers("DELETE", path) if auth and self._private_key else {}
        resp = self._session.delete(url, headers=headers, timeout=15)
        resp.raise_for_status()
        return resp.json()

    # ── Market Discovery ──────────────────────────────────────────────

    def get_events(
        self,
        status: str = "open",
        series_ticker: Optional[str] = None,
        limit: int = 200,
        cursor: Optional[str] = None,
    ) -> Dict:
        """List events, optionally filtered by series."""
        params = {"status": status, "limit": limit, "with_nested_markets": "true"}
        if series_ticker:
            params["series_ticker"] = series_ticker
        if cursor:
            params["cursor"] = cursor
        return self._get("/events", params=params)

    def get_event(self, event_ticker: str) -> Dict:
        return self._get(f"/events/{event_ticker}", params={"with_nested_markets": "true"})

    def get_markets(
        self,
        status: str = "open",
        series_ticker: Optional[str] = None,
        event_ticker: Optional[str] = None,
        limit: int = 200,
        cursor: Optional[str] = None,
    ) -> Dict:
        params = {"status": status, "limit": limit}
        if series_ticker:
            params["series_ticker"] = series_ticker
        if event_ticker:
            params["event_ticker"] = event_ticker
        if cursor:
            params["cursor"] = cursor
        return self._get("/markets", params=params)

    def get_market(self, ticker: str) -> Dict:
        return self._get(f"/markets/{ticker}")

    def get_orderbook(self, ticker: str) -> Dict:
        return self._get(f"/markets/{ticker}/orderbook")

    def get_series(self, series_ticker: str) -> Dict:
        return self._get(f"/series/{series_ticker}")

    # ── Sports Market Discovery ───────────────────────────────────────

    def find_sports_markets(self, sport_keywords: List[str]) -> List[Dict]:
        """
        Scan all open events/markets and filter for sports-related ones.
        Kalshi doesn't have a direct category filter in the API,
        so we keyword-match on titles.
        """
        all_markets = []
        cursor = None
        while True:
            data = self.get_events(status="open", limit=200, cursor=cursor)
            events = data.get("events", [])
            for event in events:
                title_lower = event.get("title", "").lower()
                sub_lower = event.get("sub_title", "").lower()
                category = event.get("category", "").lower()
                combined = f"{title_lower} {sub_lower} {category}"
                if any(kw.lower() in combined for kw in sport_keywords):
                    markets = event.get("markets", [])
                    for m in markets:
                        m["_event_title"] = event.get("title", "")
                        m["_event_category"] = event.get("category", "")
                    all_markets.extend(markets)
            cursor = data.get("cursor")
            if not cursor or not events:
                break
        return all_markets

    def find_tennis_markets(self) -> List[Dict]:
        keywords = [
            "tennis", "ATP", "WTA", "Grand Slam", "Roland Garros",
            "Wimbledon", "US Open Tennis", "Australian Open Tennis",
            "French Open", "match winner", "set winner",
        ]
        return self.find_sports_markets(keywords)

    def find_baseball_markets(self) -> List[Dict]:
        keywords = [
            "baseball", "MLB", "World Series", "American League",
            "National League", "home run", "innings", "pitcher",
            "Yankees", "Dodgers", "Mets", "Red Sox", "Cubs",
            "Astros", "Braves", "Phillies", "Padres",
        ]
        return self.find_sports_markets(keywords)

    # ── Portfolio ─────────────────────────────────────────────────────

    def get_balance(self) -> Dict:
        return self._get("/portfolio/balance", auth=True)

    def get_positions(self, settlement_status: str = "unsettled") -> Dict:
        return self._get(
            "/portfolio/positions",
            params={"settlement_status": settlement_status},
            auth=True,
        )

    def get_fills(self, limit: int = 100) -> Dict:
        return self._get("/portfolio/fills", params={"limit": limit}, auth=True)

    # ── Order Execution ───────────────────────────────────────────────

    def create_order(
        self,
        ticker: str,
        side: str,          # "yes" or "no"
        action: str,         # "buy" or "sell"
        count: int,          # number of contracts
        price_cents: int,    # limit price in cents (1-99)
        time_in_force: str = "gtc",  # "gtc", "ioc", "fill_or_kill"
    ) -> Dict:
        """Place a limit order on Kalshi."""
        order_data = {
            "ticker": ticker,
            "side": side,
            "action": action,
            "count": count,
            "type": "limit",
            "client_order_id": str(uuid.uuid4()),
            "time_in_force": time_in_force,
        }
        if side == "yes":
            order_data["yes_price"] = price_cents
        else:
            order_data["no_price"] = price_cents
        
        logger.info(f"Placing order: {action} {count}x {side} @ {price_cents}¢ on {ticker}")
        return self._post("/portfolio/orders", order_data)

    def cancel_order(self, order_id: str) -> Dict:
        return self._delete(f"/portfolio/orders/{order_id}")

    def get_orders(self, status: str = "resting") -> Dict:
        return self._get("/portfolio/orders", params={"status": status}, auth=True)


class KalshiPublicClient:
    """Unauthenticated client for public market data only."""

    BASE = "https://api.elections.kalshi.com/trade-api/v2"

    def __init__(self):
        self._session = requests.Session()

    def get_markets(self, **params) -> Dict:
        resp = self._session.get(f"{self.BASE}/markets", params=params, timeout=15)
        resp.raise_for_status()
        return resp.json()

    def get_events(self, **params) -> Dict:
        params.setdefault("with_nested_markets", "true")
        resp = self._session.get(f"{self.BASE}/events", params=params, timeout=15)
        resp.raise_for_status()
        return resp.json()

    def get_orderbook(self, ticker: str) -> Dict:
        resp = self._session.get(f"{self.BASE}/markets/{ticker}/orderbook", timeout=15)
        resp.raise_for_status()
        return resp.json()
